### PR TITLE
Deprecating ObjectConf and introducing TargetConf

### LINKED ISSUE #882: Deprecated ObjectConf for TargetConf
This is a partial implementation of #566.

## TLDR
Object configs is changing from:
```yaml
target: classname
params:
  a: 10
  b: 20
```
to:
```yaml
_target_: classname
a: 10
b: 20
```
This include Hydra's launchers and sweepers, so if they appear in your config file you will need to adjust the configs.

## Details

ObjectConf:
```python
@dataclass
class ObjectConf(Dict[str, Any]):
    # class, class method or function name
    target: str = MISSING

    # parameters to pass to cls when calling it
    params: Any = field(default_factory=dict)
```

TargetConf:
```python
@dataclass
class TargetConf:
    """
    Use by subclassing and adding fields
    e.g:
    @dataclass
    class UserConf(TargetConf):
        _target_ = "module.User"
        name: str = MISSING
        age: int = MISSING
    """

    # class, class method or function name
    _target_: str = MISSING
```

This will be deprecated for user code, but for Hydra it's a breaking change because sweeper and launchers will no longer need the pesky params to override their parameters:
```
$ python foo.py hydra.sweeper.params.max_batch_size=2
# will become
$ python foo.py hydra.sweeper.max_batch_size=2
```

### PR DESCRIPTION
**Pull request by omry**
closes #882 
- [x] Implement
- [x] Update plugins
- [x] Update docs
- [x] Clean up TODOs
- [x] Followup fix to configen, change to generate code in new structure.

### CODE REVIEW

File: website/docs/patterns/instantiate_objects/reference.md [hunk 1]
```diff
@@ -4,40 +4,65 @@ title: Reference
...
+    :return: the return value from the specified class or method
     """
+    ...
+
+# Alias for call
+instantiate = call
 ```
 
-For using these functions, the config must have a key called `target`. If a key called `params` is also present, its value is passed as keyword arguments to class/function specified by `target`.
+For using these functions, the config must have a key called `_target_`. Any additional parameters are passed as keyword arguments to class/function specified by `target`.
```
Comment (shagunsodhani) [inline website/docs/patterns/instantiate_objects/reference.md:27]: class/function specified by `target` -> class/function specified by `_target_`

File: website/docs/patterns/instantiate_objects/reference.md [hunk 2]
```diff
@@ -4,40 +4,65 @@ title: Reference
...
+    _target_: str = MISSING
+
 ```
 
+</div><div className="col col--6">
+
+```python title="Use it by subclassing and adding fields" 
+@dataclass
+class UserConf(TargetConf):
+    _target_ = "module.User"
```
Comment (shagunsodhani) [inline website/docs/patterns/instantiate_objects/reference.md:59]: `_target_: str = "module.User"`

File: website/docs/patterns/instantiate_objects/reference.md [hunk 3]
```diff
@@ -4,40 +4,65 @@ title: Reference
...
-  x: 10
+_target_: foo.Bar
+# additional parameters to pass when calling the target
+x: 10
 ```
 
 ### Example ObjectConf definition
-ObjectConf is defined in `hydra.types.ObjectConf`:
-```python
+TargetConf is defined in `hydra.types.TargetConf`:
```
Comment (shagunsodhani) [inline website/docs/patterns/instantiate_objects/reference.md:38]: `TargetConf` is defined in `hydra.types.TargetConf`:

File: hydra/_internal/utils.py [hunk 4]
```diff
@@ -601,53 +612,41 @@ def _get_kwargs(config: Union[ObjectConf, DictConfig], **kwargs: Any) -> Any:
...
+                f"Config key '{key}' is deprecated since Hydra 1.0 and will be removed in Hydra 1.1."
+                f"\nUse '_target_' instead of '{field}'.",
                 category=UserWarning,
             )
         else:
             warnings.warn(
                 f"""
-ObjectConf field '{field}' is deprecated since Hydra 1.0 and will be removed in Hydra 1.1.
-Use 'target' instead of '{field}'.""",
+    ObjectConf field '{field}' is deprecated since Hydra 1.0 and will be removed in Hydra 1.1.
```
[the remark this answers is not in the document]
Comment (omry) [reply hydra/_internal/utils.py]: In fact it could be either.

[the remark this answers is not in the document]
Comment (omry) [reply hydra/_internal/utils.py]: or none actually. look at the function signature.

File: website/docs/patterns/instantiate_objects/reference.md (hunk 2 above)
Comment (omry) [reply website/docs/patterns/instantiate_objects/reference.md:59]: Good catch. This is actually critical.
If the specified type does not match the one in the parent class it does not override properly in fucking @dataclasses.

File: website/docs/plugins/rq_launcher.md [hunk 5]
```diff
@@ -34,27 +34,26 @@ defaults:
 
 The configuration packaged with the plugin is defined [here](https://github.com/facebookresearch/hydra/blob/master/plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/config.py). The default configuration is as follows:
 
-```yaml
+```yaml title="$ python your_app.py hydra/launcher=rq --cfg hydra -p hydra.launcher"
 # @package hydra.launcher
 cls: hydra_plugins.hydra_rq_launcher.rq_launcher.RQLauncher
```
Comment (shagunsodhani) [inline website/docs/plugins/rq_launcher.md:39]: should be `_target_`

Comment (shagunsodhani) [inline website/docs/plugins/rq_launcher.md:39]: should be `_target_`

Comment (shagunsodhani) [inline website/docs/plugins/rq_launcher.md:39]: should be `_target_`

Comment (omry) [reply website/docs/plugins/rq_launcher.md]: why didn't you say so?

**Pull request merged**